lib/endpoints/class-wp-json-revisions-controller.php

```diff
@@ -0,0 +1,318 @@
+			array(
+				'methods'         => WP_JSON_Server::READABLE,
+				'callback'        => array( $this, 'get_item' ),
+				'permission_callback' => array( $this, 'get_item_permissions_check' ),
+				'args'            => array(
+					'context'          => array(
+						'default'      => 'view',
+					),
+				),
+			),
+			array(
+				'methods'         => WP_JSON_Server::DELETABLE,
+				'callback'        => array( $this, 'delete_item' ),
+				'permission_callback' => array( $this, 'delete_item_permissions_check' ),
+			))
+		);
+
+		register_json_route( 'wp', '/' . $this->parent_base . '/revisions/schema', array(
+			'methods'         => WP_JSON_Server::READABLE,
+			'callback'        => array( $this, 'get_item_schema' ),
+		) );
+
+	}
+
+	/**
+	 * Get a collection of revisions
+	 *
+	 * @param WP_JSON_Request $request Full data about the request.
+	 * @return mixed WP_Error or WP_JSON_Response.
+	 */
+	public function get_items( $request ) {
+		$revisions = wp_get_post_revisions( $request['parent_id'] );
+
+		$struct = array();
+		foreach ( $revisions as $revision ) {
+			$struct[] = $this->prepare_item_for_response( $revision, $request );
+		}
+		return $struct;
+	}
+
+	/**
+	 * Check if a given request has access to get revisions
+	 *
+	 * @param WP_JSON_Request $request Full data about the request.
+	 * @return mixed WP_Error|bool.
+	 */
+	public function get_items_permissions_check( $request ) {
+
+		$parent_id = ! empty( $request['parent_id'] ) ? $request['parent_id'] : 0;
+		$parent = get_post( $parent_id );
+		if ( ! $parent_id || ! $parent || $this->parent_post_type !== $parent->post_type ) {
+			return new WP_Error( 'json_post_invalid_parent_id', __( 'Invalid post parent ID.' ), array( 'status' => 404 ) );
+		}
+
+		$parent_post_type_obj = get_post_type_object( $parent->post_type );
+		if ( ! current_user_can( $parent_post_type_obj->cap->edit_post, $parent->ID ) ) {
+			return new WP_Error( 'json_forbidden', __( 'Sorry, you cannot view revisions of this post.' ), array( 'status' => 403 ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get one revision from the collection
+	 */
+	public function get_item( $request ) {
+		$revision = get_post( $request['id'] );
+		$response = $this->prepare_item_for_response( $revision, $request );
+		return $response;
+	}
+
+	/**
+	 * Check if a given request has access to get a specific revision
+	 *
+	 * @param WP_JSON_Request $request Full data about the request.
+	 * @return mixed WP_Error|bool.
+	 */
+	public function get_item_permissions_check( $request ) {
+
+		$response = $this->get_items_permissions_check( $request );
+		if ( ! $response || is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$post = get_post( $request['id'] );
+		if ( ! $post || 'revision' !== $post->post_type ) {
+			return new WP_Error( 'json_post_invalid_id', __( 'Invalid revision ID.' ), array( 'status' => 404 ) );
```

Comment:
Elsewhere we have not thrown invalid errors in the permissions checks. We'd typically not return any error in the case of the post not existing, or being the wrong post type - and handle that check in the `get_item` callback it's self.